### DISCUSSION TITLE
Allow passing a `jax.random.PRNGKey` to class `EngineBuilder`

### DISCUSSION THREAD

**Discussion by Seb-Lorek** [liesel-devs/liesel#152] [Ideas]
Hey `liesel` devs. Currently I'm pondering on how to safely use parallel pseudo-random number generation in `JAX` for a simulation study. In a simulation study in general it is important to have repeatable pseudo-random numbers across multiple processes (local or distributed) such that the streams are independent.  In `numpy` this can be guaranteed (with very very high prob) by [spawning](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning). In `JAX` we can either use `jax.random.split` or `jax.random.fold_in` for safe production of (pseudo) independent streams (directly or indirectly via further key derivation). However both involve passing a `jax.random.PRNGKey` and not the **seed**.  Creating a iterable/list of integers and iterating over them is not safe, corresponding to the following

```python
for i in range(n):
    rng = jax.random.PRNGKey(i)
```

See [#18211](https://github.com/google/jax/discussions/18211) and [here](https://stackoverflow.com/questions/75338838/jax-best-way-to-iterate-rngkeys) for some discussions.  

In order to reliably produce independent streams within a simulation study, it would be important to have the option of directly passing a `jax.random.PRNGKey` to the `EngineBuilder`. This should be quite easy to implement as far as I can see.

**Comment by wiep** [comment]
Passing a KeyArray was possible at some point. I do not know why we removed it. However, it should be pretty straightforward implement it in a PR. Please feel encourage to give it a try.

In the constructor of EngineBuilder, you want to check the type of the seed argument:
* If it is a KeyArray / PRNGKey use it directly in jax.split
* if it is a valid argument to PRNGKey create a new key and then split it
* Otherwise: raise error.

https://github.com/liesel-devs/liesel/blob/1ee02de035199d52448eb8788b13a720db5aafb6/liesel/goose/builder.py#L132-L133

**Comment by Seb-Lorek** [comment]
Hey sure I'll give it a try later.

**Comment by jobrachem** [comment]
Thanks for your contribution! Since the request has since been implemented in #157, I'm closing this thread.
